src/validator.ts

```diff
@@ -363,7 +407,130 @@ function _defaultValidations(): ValidationFunction[] {
       }
     }
 
-    function _assertPropertiesMatch(expected: spec.Property, actual: spec.Property, label: string, action: string) {
+    /**
+     * Check if subType is indeed covariant to superType
+     */
+    function _isCovariantOf(subType?: spec.TypeReference, superType?: spec.TypeReference): boolean {
+      // one void, while other isn't => not covariant
+      if ((subType === undefined) !== (superType === undefined)) {
+        return false;
+      }
+
+      // Same type is always covariant
+      if (deepEqual(subType, superType)) {
+        return true;
+      }
+
+      // Handle array collections (covariant)
+      if (spec.isCollectionTypeReference(subType) && spec.isCollectionTypeReference(superType)) {
+        if (subType.collection.kind === 'array' && superType.collection.kind === 'array') {
+          return _isCovariantOf(subType.collection.elementtype, superType.collection.elementtype);
+        }
+        // Maps/Records are not allowed to be covariant in C#, so we exclude them here
+        return false;
+      }
+
+      // We only support classes for named types
+      if (!spec.isNamedTypeReference(subType) || !spec.isNamedTypeReference(superType)) {
+        return false;
+      }
+
+      const subTypeSpec = _dereference(subType.fqn, assembly, validator);
+      const superTypeSpec = _dereference(superType.fqn, assembly, validator);
+
+      if (!subTypeSpec || !superTypeSpec) {
+        return false;
+      }
+
+      // Handle class-to-class inheritance
+      if (spec.isClassType(subTypeSpec) && spec.isClassType(superTypeSpec)) {
+        return _classExtendsClass(subTypeSpec, superType.fqn);
+      }
+
+      // Handle interface-to-interface inheritance
+      if (spec.isInterfaceType(subTypeSpec) && spec.isInterfaceType(superTypeSpec)) {
+        return _interfaceExtendsInterface(subTypeSpec, superType.fqn);
+      }
+
+      // Handle class implementing interface
+      if (spec.isClassType(subTypeSpec) && spec.isInterfaceType(superTypeSpec)) {
+        return _classImplementsInterface(subTypeSpec, superType.fqn);
+      }
+
```

Comment:
Needs to handle unions and intersections here as well, no? If only to reject them with an assertion?